### ISSUE TITLE
CustomScrollView repeatedly builds sliver-children that are outside of the view and cache range

### ISSUE THREAD

**Issue by ravenblackx** [flutter/flutter#92276]
When a `CustomScrollView` has multiple 'slivers' that are `SliverGrid` (or Custom), under some circumstances the ChildBuilderDelegates are called while the entire sliver is not on screen.

For example, in this test demo below, the expectation for the scrolling that's performed would be that none of the second `SliverGrid`'s children would be built at time A or B as it is not yet in the window. Instead, the first child of that sliver is built immediately, and the second child is built repeatedly while scrolling slowly moves towards where it should be built for the first time.

(Interestingly, if the first SliverGrid has `childAspectRatio: 2` instead of `childAspectRatio: 0.5`, the behavior is as it should be. `childAspectRatio: 1`, however, still misbehaves, as does setting the same size via `mainAxisExtent`.)

Increasing the number of frames during the drag (by changing timedDrag's duration) increases the number of times that second child is incorrectly rebuilt.

```
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

const _colors = [Colors.red, Colors.green, Colors.blue, Colors.yellow];

void main() {
  testWidgets('SliverGrid children should be built only when they enter the cache/view area', (tester) async {
    final builds = [0, 0, 0, 0];
    await tester.pumpWidget(
      Directionality(
        textDirection: TextDirection.ltr,
        child: Center(
          child: RepaintBoundary(
            child: SizedBox(
              width: 100,
              height: 100,
              child: CustomScrollView(
                scrollDirection: Axis.horizontal,
                cacheExtent: 0.0,
                slivers: [
                  SliverToBoxAdapter(
                    child: SizedBox(width: 20),
                  ),
                  SliverGrid(
                    delegate: SliverChildBuilderDelegate(
                      (_, index) {
                        return Container(color: _colors[index]);
                      },
                      childCount: 3,
                    ),
                    gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                      crossAxisCount: 1,
                      childAspectRatio: 0.5,
                    ),
                  ),
                  SliverGrid(
                    delegate: SliverChildBuilderDelegate(
                      (_, index) {
                        builds[index]++;
                        return Container(color: _colors[index]);
                      },
                      childCount: 4,
                    ),
                    gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                      crossAxisCount: 2,
                      childAspectRatio: 1.0,
                    ),
                  ),
                  SliverToBoxAdapter(
                    child: SizedBox(width: 20),
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
    await tester.pump();
    //await expectLater(find.byType(CustomScrollView), matchesGoldenFile('A'));
    // This expectation fails - after the view is built, the second SliverGrid
    // which is totally outside the view has built its first child.
    print('XXXXX at A: $builds');  // [1,0,0,0]
    // expect(builds, equals([0,0,0,0]));

    // Drag so the second SliverGrid is almost on-screen.
    await tester.timedDrag(find.byType(CustomScrollView), const Offset(-500, 0),
        const Duration(seconds: 4));
    //await expectLater(find.byType(CustomScrollView), matchesGoldenFile('B'));
    // This expectation fails - slowly dragging the view in a way that never
    // reveals the second SliverGrid repeatedly builds its second child.
    //expect(builds, equals([0, 0, 0, 0]));
    print('XXXXX at B: $builds');  // [1,14,0,0]

    // Drag so the second SliverGrid's first half is partially on-screen.
    await tester.timedDrag(find.byType(CustomScrollView), const Offset(-30, 0),
        const Duration(seconds: 1));
    //await expectLater(find.byType(CustomScrollView), matchesGoldenFile('C'));
    //expect(builds, equals([1, 1, 0, 0]));
    print('XXXXX at C: $builds');  // [1,15,0,0]
  });
}
```

**Comment by Piinks** [comment]
It looks like none of the Containers in your test have a height, so they all fit inside the window of your test. Therefore, they all are built.

@dnfield is right, the CustomScrollView does not build its sliver children lazily, but the Slivers themselves build their children lazily. There was a really good example recently from @craiglabenz [in this video](https://youtu.be/LUqDNnv_dh0), and [this accompanying DartPad](https://www.youtube.com/redirect?event=video_description&redir_token=QUFFLUhqbDc4bEJOLThBRHl2ME1nN3E0RkIyNHFUdDJuQXxBQ3Jtc0ttXzJ0NXMtS2xDM0FYZjlGUjRmaWdFRjJoVWxzdzdwZG0wTkpkMnRGVnBlcGJNc2o3T3ZLUmZxSDRTNW5XSDV5cnNGMzRMNzVmNDBCMk1QRFVSUzd0eTNjVEVSeHgxQUtfM1JRaFFWUlBjcVg2OGljbw&q=https%3A%2F%2Fgoo.gle%2F3oCRRTt) that shows multiple SliverLists in a CustomScrollView building all of the children lazily as expected.

**Comment by craiglabenz** [comment]
I would note that I learned that "really good example" code from @Piinks, the Queen of Scrolling 😆

**Comment by ravenblackx** [comment]
> It looks like none of the Containers in your test have a height, so they all fit inside the window of your test. Therefore, they all are built.

It's a horizontal scrolling `CustomScrollView`, so the lack of explicit height isn't relevant. :)

Also, if the issue was that they were all on screen in the first place, then the number of builds in the test should have been [1,1,1,1] not [1,15,0,0], so it would still be doing the wrong thing for that case.

**Comment by Piinks** [comment]
Oh interesting! Thanks for pointing that out. I had not noticed. There isn't a width either, so I'm not really sure what would be lazily built if it all fits in the viewport.

**Comment by Piinks** [comment]
I've put this into an app and run it to see what happens to be sure. I don't see the extra build calls happening when scrolling, it appears to be working as intended..
This is the app I put together:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MyWidget();
  }
}

class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: RepaintBoundary(
            child: SizedBox(
              width: 100,
              height: 100,
              child: CustomScrollView(
                scrollDirection: Axis.horizontal,
                cacheExtent: 0.0,
                slivers: [
                  SliverToBoxAdapter(
                    child: SizedBox(width: 20),
                  ),
                  SliverGrid(
                    delegate: SliverChildBuilderDelegate(
                      (_, index) {
                        return Container(color: Colors.red);
                      },
                      childCount: 3,
                    ),
                    gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                      crossAxisCount: 1,
                      childAspectRatio: 0.5,
                    ),
                  ),
                  SliverGrid(
                    delegate: SliverChildBuilderDelegate(
                      (_, index) {
                        print('Index $index built');
                        return Container(color: Colors.blue);
                      },
                      childCount: 4,
                    ),
                    gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                      crossAxisCount: 2,
                      childAspectRatio: 1.0,
                    ),
                  ),
                  SliverToBoxAdapter(
                    child: SizedBox(width: 20),
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

If you are able to reproduce the extra build calls in an app, please update and we'll take another look! Thanks!

**Comment by ravenblackx** [comment]
I don't work with Flutter any more, so I don't have the environment set up for it. I tried to run your example app in flutlab.io but it wouldn't scroll at all there. Running my test in flutlab.io still outputs the numbers showing it built more times than it should have, though, and I did initially encountered the behavior in a non-test context, so it wasn't limited to test-only behavior. Is a test case demonstrating the malbehavior not good enough?

flutlab.io test case output from pasting in the exact content in the original report:
```
⭐ Build started at Wednesday, December 1st 2021, 5:30:37PM -08:00 ⭐ 
 🏁 Build finished at Wednesday, December 1st 2021, 5:31:00PM -08:00 🏁 
 ✅ Build completed successfully. 
00:15 +0: SliverGrid children should be built only when they enter the cache/view area                                                                                                                 
00:15 +0: SliverGrid children should be built only when they enter the cache/view area                                                                                                                 
XXXXX at A: [1, 0, 0, 0]
XXXXX at B: [1, 14, 0, 0]
XXXXX at C: [1, 15, 0, 0]

00:15 +1: SliverGrid children should be built only when they enter the cache/view area                                                                                                                 
00:15 +1: All tests passed!         
```

(Bearing in mind the test passes because I made the output prints rather than failures so it could run to the end.)

If the test case doesn't reproduce the excessive building behavior in your environment then maybe it has been fixed in some version later than whatever I was using and whatever flutlab.io is using.

But as I say, I'm no longer invested in it working correctly other than a vague, general dislike of things not working correctly, so I'm not going to be making a big effort to prove that there's an issue. If that means it gets closed not reproducible, okay, I guess. :)

**Comment by Piinks** [comment]
I think I came to understand the issue here in https://github.com/flutter/flutter/issues/95028

In a CustomScrollView that contains multiple SliverLists/Grids, at least the first child in each one is built. This is because we require at least one child to estimate the size of the viewport. I closed #95028 as WAI in this case, and instead filed https://github.com/flutter/flutter/issues/97676 to examine if there are opportunities to improve scroll extent estimates, whether that be through adding other options or removing the need to have at least one child built for every sliver in order to make that calculation.

I think that covers this, so I am going to optimistically close this. :) If not, please do feel free to reopen and we can examine further.
